engine/classes/Elgg/Filesystem/Directory.php

```diff
@@ -0,0 +1,80 @@
+<?php
+namespace Elgg\Filesystem;
+
+/**
+ * A simple directory abstraction.
+ * 
+ * @since 1.10.2
+ * 
+ * @access private
+ */
+interface Directory {
+
+	/**
+	 * Returns a subdirectory with access limited to the given directory.
+	 * 
+	 * @param string $path The path relative to this directory to chroot to.
+	 * 
+	 * @return Directory A new directory instance.
+	 */
+	public function chroot($path);
+
+	/**
+	 * Read the file off the filesystem.
+	 * 
+	 * @param string $path The directory-relative path to the target file.
+	 * 
+	 * @return string Empty string if the file doesn't exist.
+	 */
+	public function getContents($path);
+	
+	/**
+	 * A reference to the file at the given path, even if it doesn't exist yet.
+	 * 
+	 * However, will throw an exception if the file is already a directory.
+	 * 
+	 * @param string $path The path to the file, relative to this directory.
+	 * 
+	 * @return File
+	 */
+	public function getFile($path);
+	
+	/**
+	 * Recursively list the files in the given directory path.
+	 * 
+	 * @param string $path The subdirectory path within this directory
+	 * 
+	 * @return Collection<File>
```

Comment:
Collection is not included in this file anywhere, and I have no idea if any annotation parser has support for generics, but I figured I'd throw it in for human readability's sake. If this causes problems I have no problems tweaking it.